cert/cert.go

```diff
@@ -0,0 +1,225 @@
+	return err
+}
+
+// NewCA generates a CA certificate/key pair suitable for signing server
+// keys for an environment with the given name.
+func NewCA(envName, UUID string, expiry time.Time) (certPEM, keyPEM string, err error) {
+	key, err := rsa.GenerateKey(rand.Reader, KeyBits)
+	if err != nil {
+		return "", "", err
+	}
+	// TODO(perrito666) 2016-05-02 lp:1558657
+	now := time.Now()
+
+	serialNumber, err := newSerialNumber()
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+
+	template := &x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			CommonName:   fmt.Sprintf("juju-generated CA for model %q", envName),
+			Organization: []string{"juju"},
+			SerialNumber: UUID,
+		},
+		NotBefore:             now.UTC().AddDate(0, 0, -7),
+		NotAfter:              expiry.UTC(),
+		SubjectKeyId:          bigIntHash(key.N),
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		IsCA:                  true,
+		MaxPathLen:            0, // Disallow delegation for now.
+		BasicConstraintsValid: true,
+	}
+	certDER, err := x509.CreateCertificate(rand.Reader, template, template, &key.PublicKey, key)
+	if err != nil {
+		return "", "", fmt.Errorf("cannot create certificate: %v", err)
+	}
+	certPEMData := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certDER,
+	})
+	keyPEMData := pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(key),
+	})
+	return string(certPEMData), string(keyPEMData), nil
+}
+
+// newSerialNumber returns a new random serial number suitable
+// for use in a certificate.
+func newSerialNumber() (*big.Int, error) {
+	// A serial number can be up to 20 octets in size.
+	// https://tools.ietf.org/html/rfc5280#section-4.1.2.2
+	n, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 8*20))
+	if err != nil {
+		return nil, errors.Annotatef(err, "failed to generate serial number")
+	}
+	return n, nil
+}
+
+// NewDefaultServer generates a certificate/key pair suitable for use by a server, with an
+// expiry time of 10 years.
+func NewDefaultServer(caCertPEM, caKeyPEM string, hostnames []string) (certPEM, keyPEM string, err error) {
+	// TODO(perrito666) 2016-05-02 lp:1558657
+	expiry := time.Now().UTC().AddDate(10, 0, 0)
+	return newLeaf(caCertPEM, caKeyPEM, expiry, hostnames, []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth})
+}
+
+// NewServer generates a certificate/key pair suitable for use by a server.
+func NewServer(caCertPEM, caKeyPEM string, expiry time.Time, hostnames []string) (certPEM, keyPEM string, err error) {
+	return newLeaf(caCertPEM, caKeyPEM, expiry, hostnames, []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth})
+}
+
+// newLeaf generates a certificate/key pair suitable for use by a leaf node.
+func newLeaf(caCertPEM, caKeyPEM string, expiry time.Time, hostnames []string, extKeyUsage []x509.ExtKeyUsage) (certPEM, keyPEM string, err error) {
```

Comment:
as above, I think this should take CommonName as an arg, and juju/cert can have its own version that passes in "*"